### PR TITLE
[APIS-782] JDBC isValid() in CUBRIDConnection throws UnsupportedOperationException

### PR DESCRIPTION
**Pull request by byungwook-kim**
http://jira.cubrid.org/browse/APIS-782

### CODE REVIEW

**Comment by cubridci** [comment]
Build comment file: 
--------------
**
** This Pull Request's title is '[APIS-782] JDBC isValid() in CUBRIDConnection throws UnsupportedOperationException'.
** And it does not start with a related JIRA issue number **
** Please edit this Pull Request's title. (eg: '[CBRD-20309] Redesign vacuum data flush' **
**

--------------

**Comment by cubridci** [comment]
Build comment file: 
--------------
**
** This Pull Request's title is '[APIS-782] JDBC isValid() in CUBRIDConnection throws UnsupportedOperationException'.
** And it does not start with a related JIRA issue number **
** Please edit this Pull Request's title. (eg: '[CBRD-20309] Redesign vacuum data flush' **
**

--------------

File: src/jdbc/cubrid/jdbc/net/BrokerHandler.java [hunk 1]
```diff
@@ -227,6 +228,20 @@ public static int statusBroker(String ip, int port, int process, byte[] session,
...
+    public static boolean isValid(String ip, int port, int process, byte[] session, int timeout)
+        throws SQLException {
+
+        try {
+            if (BrokerHandler.statusBroker(ip, port, process, session, timeout) == -2) {
+                return false;
+            }
+        } catch (Exception e) {
+            throw new SQLException("Invalid Connection");
+        }
```
Comment (hgryoo) [inline src/jdbc/cubrid/jdbc/net/BrokerHandler.java:240]: According to JDBC Spec, SQLException should be thrown **if the value supplied for timeout is less then 0**.

File: src/jdbc/cubrid/jdbc/net/BrokerHandler.java [hunk 2]
```diff
@@ -227,6 +228,20 @@ public static int statusBroker(String ip, int port, int process, byte[] session,
         return status;
     }
 
+    public static boolean isValid(String ip, int port, int process, byte[] session, int timeout)
```
Comment (hgryoo) [inline src/jdbc/cubrid/jdbc/net/BrokerHandler.java:231]: This function just wraps BrokerHandler.statusBroker(). I think we don't need to declare isValid () in the BrokerHandler class. It looks not related to other functions in the same class.

File: src/jdbc/cubrid/jdbc/jci/UConnection.java [hunk 3]
```diff
@@ -1832,6 +1833,15 @@ UInputBuffer send_recv_msg() throws UJciException, IOException {
 		return send_recv_msg(true);
 	}
 
+	public boolean isValid(int timeout) throws SQLException {
+		boolean valid = !isClosed;
+
+                if (protoVersionIsAbove(PROTOCOL_V9)) {
```
Comment (hgryoo) [inline src/jdbc/cubrid/jdbc/jci/UConnection.java:1839]: please indent

File: src/jdbc/cubrid/jdbc/driver/CUBRIDConnection.java [hunk 4]
```diff
@@ -988,8 +988,14 @@ public String getClientInfo(String arg0) throws SQLException {
...
-	public boolean isValid(int arg0) throws SQLException {
-		throw new SQLException(new java.lang.UnsupportedOperationException());
+	public boolean isValid(int timeout) throws SQLException {
+		if (u_con == null || is_closed) return false;
+
+		try {
+			return u_con.isValid(timeout);
+		} catch (SQLException e) {
+			return false;
+		}
```
Comment (hgryoo) [inline src/jdbc/cubrid/jdbc/driver/CUBRIDConnection.java:998]: According to JDBC Spec, SQLException should be thrown if the value supplied for timeout is less then 0.

File: src/broker/broker.c [hunk 5]
```diff
@@ -849,7 +849,7 @@ receiver_thr_f (void *arg)
 
       if (strncmp (cas_req_header, "ST", 2) == 0)
 	{
-	  int status = -1;
+	  int status = -2;
```
Comment (hgryoo) [inline src/broker/broker.c:852]: The meaning of initialized value (-2) is not clear. It should be defined as a enum value (in cas_function.h) such as `FN_STATUS_NONE = -2`.

**Comment by cubridci** [comment]
Build comment file: 
--------------
#### File: src/broker/cas_common.h (5 incorrect indentation(s) found)
```
   162   FN_STATUS_NONE = -2,
   163   FN_STATUS_IDLE = -1,
   164   FN_STATUS_CONN = 0,
   165   FN_STATUS_BUSY = 1,
   166   FN_STATUS_DONE = 2
```
--------------
**Please run 'indent -l120 -lc120' to correct the indentation**
**
** This Pull Request's title is '[APIS-782] JDBC isValid() in CUBRIDConnection throws UnsupportedOperationException'.
** And it does not start with a related JIRA issue number **
** Please edit this Pull Request's title. (eg: '[CBRD-20309] Redesign vacuum data flush' **
**

--------------

File: src/jdbc/cubrid/jdbc/jci/UConnection.java [hunk 6]
```diff
@@ -1835,11 +1841,16 @@ UInputBuffer send_recv_msg() throws UJciException, IOException {
...
-		}
-		return valid;
+		try {
+			int status = BrokerHandler.statusBroker(CASIp, CASPort, processId, sessionId, timeout);
+			if (status == UConnection.FN_STATUS_NONE) {
+    				return false;
+    	    		}
+    		} catch (Exception e) {
+    			return false;
+    		}
```
Comment (hgryoo) [inline src/jdbc/cubrid/jdbc/jci/UConnection.java:1851]: Protocol version checking seems to be missing. `protoVersionIsAbove (PROTOCOL_V9)`

**Comment by cubridci** [comment]
Build comment file: 
--------------
#### File: src/broker/cas_common.h (5 incorrect indentation(s) found)
```
   162   FN_STATUS_NONE = -2,
   163   FN_STATUS_IDLE = -1,
   164   FN_STATUS_CONN = 0,
   165   FN_STATUS_BUSY = 1,
   166   FN_STATUS_DONE = 2
```
--------------
**Please run 'indent -l120 -lc120' to correct the indentation**
**
** This Pull Request's title is '[APIS-782] JDBC isValid() in CUBRIDConnection throws UnsupportedOperationException'.
** And it does not start with a related JIRA issue number **
** Please edit this Pull Request's title. (eg: '[CBRD-20309] Redesign vacuum data flush' **
**

--------------

**Comment by cubridci** [comment]
Build comment file: 
--------------
**
** This Pull Request's title is '[APIS-782] JDBC isValid() in CUBRIDConnection throws UnsupportedOperationException'.
** And it does not start with a related JIRA issue number **
** Please edit this Pull Request's title. (eg: '[CBRD-20309] Redesign vacuum data flush' **
**

--------------
